### PR TITLE
feat: add NodeJS version select + use @discordjs/builders + better types

### PR DESCRIPTION
**Pull request by Favna**
- Use @discordjs/builders wherever applicable for markdown output.
- Add types for MDN
- Add types for Node
- Add "version" option to Node which uses options to select between 12, 14 and 16 (default)
- Changed the docs Algolia query to send API app/key with headers

This also means this PR supersedes #19. Sorry for stealing your PR @KrishAgarwal2811 but I had this planned for a while... lol

### CODE REVIEW

**Review by iCrawl — APPROVED** [review]

File: src/functions/guide.ts [hunk 1]
```diff
@@ -15,38 +16,41 @@ export async function djsGuide(
...
+				query: search,
+			}),
 		}),
-	}).then((res) => res.json());
+		headers: {
+			'Content-Type': 'application/json',
+			'X-Algolia-API-Key': ALGOLIA_KEY,
+			'X-Algolia-Application-Id': ALGOLIA_APP,
+		},
+	}).then((res) => res.json() as Promise<AlgoliaSearchResult>);
```
Comment (NotSugden) [inline src/functions/guide.ts:32]:
```suggestion
	}).then((res) => res.json()) as AlgoliaSearchResult;
```

File: src/functions/mdn.ts [hunk 2]
```diff
@@ -1,42 +1,46 @@
...
+const cache = new Map<string, Document>();
 
 export async function mdnSearch(res: Response, query: string, target?: string): Promise<Response> {
 	query = query.trim();
 	try {
 		const qString = `${API_BASE_MDN}/api/v1/search?${encode({ q: query })}`;
 		let hit = cache.get(qString);
 		if (!hit) {
-			const result = await fetch(qString).then((r) => r.json());
+			const result = await fetch(qString).then((r) => r.json() as Promise<MdnAPI>);
```
Comment (NotSugden) [inline src/functions/mdn.ts:17]:
```suggestion
			const result = await fetch(qString).then((r) => r.json()) as MdnAPI;
```

File: src/functions/node.ts [hunk 3]
```diff
@@ -64,38 +64,59 @@ function findResult(data: any, query: string) {
...
 	query = query.trim();
 	try {
-		if (!data) {
-			data = await fetch(`${API_DOCS_NODE}/all.json`).then((r) => r.json());
+		const url = `${API_DOCS_NODE}/dist/${version}/docs/api/all.json`;
+		let allNodeData = cache.get(url);
+
+		if (!allNodeData) {
+			// Get the data for this version
+			const data = await fetch(url).then((r) => r.json() as Promise<NodeDocs>);
```
Comment (NotSugden) [inline src/functions/node.ts:82]:
```suggestion
			const data = await fetch(url).then((r) => r.json()) as NodeDocs;
```

**Review by almostSouji — CHANGES_REQUESTED** [review]
needs a rebase

**Review by almostSouji — CHANGES_REQUESTED** [review]
option description nitpick, lgtm otherwise
will probably wait for said builders release before merging

**Review by almostSouji — APPROVED** [review]

**Comment by KrishAgarwal2811** [comment]
Wow that's a great change @Favna. I just had a temporary fix 😂.

**Pull request merged**